### PR TITLE
Added juliapkg as a package

### PR DESCRIPTION
**Pull request by thewchan**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by thewchan** [comment]
@conda-forge/staged-recipes ready for review!

**Comment by isuruf** [comment]
cc @ngam

**Comment by thewchan** [comment]
@ngam I'm only adding this because it's a dependency for https://github.com/conda-forge/staged-recipes/pull/20379

**Comment by thewchan** [comment]
@ngam if you think these 2 packages are not appropriate for Conda Forge, I'll close the PR.

**Comment by thewchan** [comment]
Understood, let me know if you need anything from me.

**Comment by mkitti** [comment]
This is technically a pure Python package. The other package juliacall, which depends on this, is a replacement for pyjulia.

The main question is if this is installed with the julia-feedstock, can it detect julia there? Since julia installed by conda-forge should be on the PATH, I believe it should.

cc: @cjdoris

**Comment by cjdoris** [comment]
@thewchan Thanks so much for adding this recipe, and the one for JuliaCall! I have been meaning to get around to it myself...

For the name, how about pyjuliapkg (and pyjuliacall for JuliaCall)? The "py" prefix seems to be the consensus for Python packages (pyjulia, pytorch, pyqt, etc). That should make it especially clear that this is not a Julia package - given there has already been some confusion about this above.

The recipe as it is looks great. If the user also installs Julia from Conda (and is a compatible version), then it will be picked up and used by JuliaPkg/JuliaCall automatically simply because it is in the PATH. Though note that if JuliaUp is also installed, JuliaPkg will use that instead of whatever is in the PATH - perhaps JuliaPkg could change its preference order in this case, something to think about,

I don't think it has been suggested yet, but I'd be against making the Julia package a dependency of JuliaPkg/JuliaCall, because it frees the user to use a different Julia installation if they wish.

Would you mind making me a recipe maintainer too?

**Comment by cjdoris** [comment]
@ngam I'm not sure what exactly you are suggesting, can you be more specific please?

Just to lay out how these packages interact:
- JuliaPkg is a Python library which will find/install Julia plus any Julia packages required. The packages are installed into an environment-specific project, so if you have several virtual environments or Conda environments using JuliaPkg, each will get its own Julia project.
- JuliaCall is a Python library for calling Julia functions from Python. It depends on JuliaPkg to install Julia itself. Most of its implementation is in the Julia package PythonCall, hence JuliaCall uses JuliaPkg to ensure PythonCall is installed too (by declaring a dependency in `juliapkg.json`).
- PythonCall is a Julia package for calling Python functions from Julia. It also handles the case when Julia is being embedded into Python using JuliaPkg. PythonCall and JuliaCall are really two sides of one package, they cannot be used independently of each other, hence they are hosted in the same repo, and are released in lock-step with the same version number.

**Comment by mkitti** [comment]
If juliaup is installed then there probably should be julia on the path. I'm not sure if I checked where juliaup installs itself.

Unless we modified something, it might be in `~/.julia`

**Comment by cjdoris** [comment]
Feel free to open an issue on the JuliaPkg repo to discuss the order to search for Julia. It seems off-topic for the present PR.

**Comment by mkitti** [comment]
If we have juliaup installed in $JULIA_DEPOT_PATH will it find juliaup there?

https://github.com/conda-forge/julia-feedstock/blob/main/recipe/scripts/activate.sh

**Comment by cjdoris** [comment]
> If we have juliaup installed in $JULIA_DEPOT_PATH will it find juliaup there?

No, right now JuliaPkg only looks for JuliaUp in the PATH. That could be changed of course.

**Comment by mkitti** [comment]
> Nevertheless, JuliaPkg and JuliaCall are functioning Python packages which can deal with their own Julia dependencies, so distributing them as-is via Conda is fine, no? I appreciate they can be better integrated with other Conda packages that may provide Julia dependencies, but that can be a job for later when a general strategy for distributing Julia packages via Conda is established.

My understanding is "no". The current situation of certain Python packages managing their own Julia dependencies is meant to be temporary because we only have a handful of packages, and we frankly did not have a solution to install Julia packages.

https://conda-forge.org/docs/user/introduction.html
> Common standards ensure that all packages have compatible versions.

By building the depots at build time conda-forge can ensure there is a compatible package set. If the versions are determined on the user's machine, then conda-forge cannot offer any guarantees of compatibility

To be clear I'm distinguishing conda from conda-forge here. I do not speak for conda-forge, but this my understanding of what differentiates a conda-forge package from any other conda package.

**Comment by cjdoris** [comment]
I think this package and the JuliaCall one can both be merged pretty much as-is. As the generic framework for distributing Julia packages via Conda develops, we can hook into that (it should be as simple as setting an env var), but I don't think we're at that stage yet.

The main question mark is about the name. I think `pyjuliapkg` and `pyjuliacall`. Opinions?

**Comment by mkitti** [comment]
I also think it should be pyjuliapkg and pyjuliacall.

There is also an R package called JuliaCall:
https://cran.r-project.org/web/packages/JuliaCall/index.html
https://github.com/Non-Contradiction/JuliaCall

That is older and completely unrelated to what is being packaged here.

What is being packaged here is
https://pypi.org/project/juliapkg/
https://pypi.org/project/juliacall/

As I mentioned on the pysr pull request, we can use these tools within the build phase similar to how we used pyjulia in the build phase for pysr.

**Comment by mkitti** [comment]
Also there is no JuliaCall.jl. That package is PythonCall.jl and would presumably be called julia-pythoncall in conda-forge

**Comment by mkitti** [comment]
I'm heavily leaning towards the name pyjuliapkg because the git repository is https://github.com/cjdoris/pyjuliapkg

**Comment by mkitti** [comment]
As is still uses the wrong name I think

**Comment by cjdoris** [comment]
Call it `pyjuliapkg`. To take an example, `torch` on PyPI is called `pytorch` in conda-forge. This is because `pytorch` is the Python wrapper of a C++ library called `torch` (also in conda-forge). So there is a precedent here for using the "py" prefix to disambiguate Python packages from other related packages within conda-forge. And the names can differ from PyPI because there is no such ambiguity there - everything on PyPI is a python package.

**Comment by cjdoris** [comment]
As I believe you already worked out, `julia-pkg` would be a different thing, namely the Julia package `Pkg` itself, bundled with the strategy used over on `pysr` etc.

**Comment by cjdoris** [comment]
Regarding integrating JuliaPkg with the proposed strategy for installing Julia packages via conda-forge:

[JuliaPkg can be configured](https://github.com/cjdoris/pyjuliapkg#configuration) to work offline using a given Julia executable and given Julia project by setting these env vars:
```bash
PYTHON_JULIAPKG_OFFLINE=yes
PYTHON_JULIAPKG_EXE=/path/to/julia/executable
PYTHON_JULIAPKG_PROJECT=/path/to/julia/project
```

These lines can be added to the activate script for the Julia package in conda-forge, but should only be done once it is possible to install all the required Julia dependencies via conda-forge.

Most importantly, all these dependencies must be in a single project. As far as I understand, the current strategy installs Julia packages into separate Julia projects and there is no automatic mechanism to put them all into a single project.

File: recipes/juliapkg/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  host:
+    - pip
+    - python >=3
+  run:
+    - python >=3
+    - semantic_version >=2.9,<3.dev0
+
+test:
+  imports:
+    - juliapkg
```
Comment (mkitti) [inline recipes/juliapkg/meta.yaml:27]:
```suggestion
    - pyjuliapkg
```

**Review by mkitti — CHANGES_REQUESTED** [review]

**Comment by mkitti** [comment]
Also I think we should add @cjdoris as a maintainer of the feedstock if he would like.

File: recipes/juliapkg/meta.yaml (hunk 1 above)
Comment (mkitti) [reply recipes/juliapkg/meta.yaml:27]: I see

**Comment by cjdoris** [comment]
> Also I think we should add @cjdoris as a maintainer of the feedstock if he would like.

Yes please.

**Comment by mkitti** [comment]
As far as I'm concerned this is a Python package. It also happens to have some code to download Julia and Julia packages. We might need to do some work to make everything play nice together down the road.

**Comment by ocefpaf** [comment]
> @ocefpaf or @isuruf to have one additional look since they've been involved in the discussions.

I'll try to look at this over the weekend but at a glance it seems OK to me.

**Comment by thewchan** [comment]
@ocefpaf @isuruf pinging to ensure this isn't buried

**Comment by mkitti** [comment]
I'm going to reiterate that I strongly feel this package should be "pyjuliapkg" named after the git repository rather than "juliapkg" as it is on PyPI. Outside of the context PyPI, the name "juliapkg" would be extremely confusing here on conda-forge, which I'm told is meant for more languages than Python.

Without that name change, I cannot offer my endorsement.

**Comment by ocefpaf** [comment]
> I strongly feel this package should be "pyjuliapkg" named after the git repository rather than "juliapkg" as it is on PyPI

This is always problematic. We want to have the original name but we also want to reflect what PyPI users migrating elsewhere will expect. I don't know the reasons why the package has a different name on PyPI but here we, to make this a bit easier for users, to use multiple outputs and reserve both names for this package.

**Comment by mkitti** [comment]
> This is always problematic. We want to have the original name but we also want to reflect what PyPI users migrating elsewhere will expect. I don't know the reasons why the package has a different name on PyPI but here we, to make this a bit easier for users, to use multiple outputs and reserve both names for this package.

The name makes perfect sense within PyPI. Within the upstream `pyjuliapkg` repository, the python package is called [`juliapkg`](https://github.com/cjdoris/pyjuliapkg/tree/main/src/juliapkg).

Within a Conda where I'm still trying to make slow progress on packaging Julia, I think "juliapkg" could be very easily confused for another piece of software, particularly one that directly involves Julia rather than Python.

1. https://github.com/JuliaLang/Pkg.jl
2. https://github.com/ararslan/julia-pkg

xref:
https://github.com/conda-forge/conda-forge.github.io/issues/18
https://github.com/conda-forge/conda-forge.github.io/issues/1822

**Pull request closed without merging**